✨[#5] 회원 가입

ramleeramlee opened the pull request:
## #️⃣연관된 이슈
- 이슈 번호: #5 

## 작업 내용
- 회원 가입 기능 구현
- dependencies 추가

## 체크리스트
- [x] 코드가 제대로 동작하는지 확인했습니다.
- [x] 문서(코드, 주석, README 등)를 업데이트했습니다.

ramleeramlee committed:
➕ [#5] dependencies 추가

- security 주석 해제
- jwt 버전 0.12.5 -> 0.11.5 변경

ramleeramlee committed:
✨ [#5] 회원가입, 로그인 기능 구현

je-pa commented on goodsending/src/main/java/com/goodsending/member/entity/Member.java:
```diff
@@ -0,0 +1,50 @@
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  @Column(name = "member_id")
+  private Long member_id;
```
자바 변수는 카멜케이스로 맞춰주시면 좋을 것 같습니다!

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/security/MemberDetailsImpl.java:
```diff
@@ -0,0 +1,65 @@
+
+public class MemberDetailsImpl implements UserDetails {
+
+  private final Member member;
```
Member 엔티티를 바로 사용하시는 것보다 필요한것만 분리하시면 좋을 것 같아요!

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/security/MemberDetailsImpl.java:
```diff
@@ -0,0 +1,65 @@
+
+  private final Member member;
+
+  public MemberDetailsImpl(Member member) {
```
lombok 사용하는걸로 바꾸셔도 좋을 것 같아요!

ramleeramlee replied:
수정 했습니다.

puclpu commented on goodsending/src/main/java/com/goodsending/member/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,85 @@
+            .permitAll() // resources 접근 허용 설정
+            //.requestMatchers("/").permitAll() // 메인 페이지 요청 허가
+            .requestMatchers("/api/members/signup")
+            .permitAll() // '/api/members/'로 시작하는 요청 모두 접근 허가
```
주석에는 "/api/members/"로 시작하는 요청 모두 접근 허가라고 하셨는데, 위 코드에서는 "/api/members/signup" 요청만 접근 허가하는 코드같습니다.

그리고 접근 허가는 위에서 부터 아래로, 좁은 범위에서 넓은 범위로 허가해주어야 하기 때문에, 만약 메인 페이지 요청 접근을 허가한다면 members 페이지 접근 허가 코드가 메인 페이지 접근 허가 코드 아래에 적혀야 할 것 같습니다.

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+    // 이메일 중복 확인
+    Optional<Member> checkEmail = memberRepository.findByEmail(email);
+    if (checkEmail.isPresent()) {
+      throw new IllegalArgumentException("중복된 Email 입니다.");
```
저희 CustomException 맞춰서 사용해주시면 좋을 것 같아요!

je-pa commented on goodsending/src/main/java/com/goodsending/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+
+    // 사용자 ROLE 확인
+    MemberRole role = MemberRole.USER;
+//        if (signupRequestDto.isAdmin()) {
```
요런거 후에 commit하실 때 위에 // TODO: {해야하는 내용} 언급해주시면 좋을 것 같아요!

ramleeramlee replied:
수정 했습니다.

puclpu commented on goodsending/src/main/java/com/goodsending/member/controller/MemberController.java:
```diff
@@ -0,0 +1,42 @@
+  @PostMapping("/members/signup")
+  public ResponseEntity<?> signup(@RequestBody @Valid SignupRequestDto signupRequestDto,
+      BindingResult bindingResult) {
+    // Validation 예외처리
```
Validation 예외처리는 global 패키지의 ExceptionHandler의   MethodArgumentNotValidException 메서드에서 처리가 가능할 것 같습니다.

ramleeramlee replied:
수정 했습니다.

puclpu commented on goodsending/src/main/java/com/goodsending/member/dto/SignupRequestDto.java:
```diff
@@ -0,0 +1,24 @@
+@Getter
+public class SignupRequestDto {
+
+  @NotBlank
```
Bean Validation에 메시지를 추가하는 것도 좋을 것 같습니다

@NotBlank(message = "이메일이 입력되지 않았습니다.") 처럼 입력값이 어떻게 유효성 검증에 실패했는지 사용자에게 알려줄 수 있습니다.

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+
+  // 회원 가입
+  public ResponseEntity<?> signup(SignupRequestDto signupRequestDto) {
+    String email = signupRequestDto.getEmail();
+    String password = passwordEncoder.encode(signupRequestDto.getPassword());
+    String phone_number = signupRequestDto.getPhone_number();
```
Member에서 static 메서드로 SignupRequestDto자체를 받는 메서드를 생성하셔서 작업하시면 서비스 단이 좀 더 깔끔해 질 것 같아요!

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/util/JwtUtil.java:
```diff
@@ -0,0 +1,89 @@
+    try {
+      Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
+      return true;
+    } catch (SecurityException | MalformedJwtException | SignatureException e) {
+      log.error("Invalid JWT signature, 유효하지 않는 JWT 서명 입니다.");
+    } catch (ExpiredJwtException e) {
+      log.error("Expired JWT token, 만료된 JWT token 입니다.");
+    } catch (UnsupportedJwtException e) {
+      log.error("Unsupported JWT token, 지원되지 않는 JWT 토큰 입니다.");
+    } catch (IllegalArgumentException e) {
+      log.error("JWT claims is empty, 잘못된 JWT 토큰 입니다.");
```
요부분 후에 Exception 처리 하시는 것도 좋을 것 같아요!

je-pa reviewed (commented):
고생하셨습니다. 추가로 MemberId를 가져올 수 있는 어노테이션하나 만들어주시면 팀원들이 작업하시기 좋으실 것 같아요!

puclpu reviewed (commented):
수고하셨습니다!

puclpu commented on goodsending/src/main/java/com/goodsending/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+//        }
+
+    // 사용자 등록
+    Member member = new Member(email, password, phone_number, role);
```
new를 통해 객체를 생성하는 것 보다 Builder와 정적 팩토리 메서드를 찾아보시고 적용해보시는 것을 추천드립니다

ramleeramlee replied:
수정 했습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/member/controller/MemberController.java:
```diff
@@ -0,0 +1,42 @@
+  private final MemberService memberService;
+
+  // 회원 가입
+  @PostMapping("/members/signup")
```
저희 swagger 사용할 예정이기 때문에 아래 예시처럼 설명 추가해주시면 좋을 것 같아요!
@Operation(summary = "선택한 강의 좋아요 토글 기능", description = "인증이 된 사용자가 선택한 강의에 좋아요 버튼을 누른다.")

ramleeramlee replied:
수정 했습니다.

je-pa reviewed (commented):
추가 리뷰 드렸습니다.

ramleeramlee committed:
♻️ [#5] 코드 피드백 수정

- 카멜케이스 적용
- MemberId 어노테이션 추가
- TODO 표시
- MemberDetailsDto 추가
- 접근 허가 주소 수정
- validation 예외처리 삭제
- Bean Validation 메시지 추가
- new 객체 -> Builder로 수정
- controller에 swagger 설명 추가

ramleeramlee committed:
♻️ [#5] 코드 피드백 수정2

- 접근 허가 주소 수정
- 메서드 이름 수정
- MemberDetailsDto memberId 추가
- MemberDetailsImpl getMemberId() 메서드 추가
